passes/pmgen/ap3_wrapcarry.cc

```diff
@@ -0,0 +1,158 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2020 QuickLogic Corp.
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/sigtools.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+#include "passes/pmgen/ap3_wrapcarry_pm.h"
+
+void create_ap3_wrapcarry(ap3_wrapcarry_pm &pm)
+{
+	auto &st = pm.st_ap3_wrapcarry;
+
+#if 0
+	log("\n");
+	log("carry: %s\n", log_id(st.carry, "--"));
+	log("lut:   %s\n", log_id(st.lut, "--"));
+#endif
+
+	log("  replacing LUT4 + QL_CARRY with $__AP3_CARRY_WRAPPER cell.\n");
+
+	Cell *cell = pm.module->addCell(NEW_ID, ID($__AP3_CARRY_WRAPPER));
+	pm.module->swap_names(cell, st.carry);
+
+	cell->setPort(ID::A, st.carry->getPort(ID(I0)));
+	cell->setPort(ID::B, st.carry->getPort(ID(I1)));
+	auto CI = st.carry->getPort(ID::CI);
+	cell->setPort(ID::CI, CI);
+	cell->setPort(ID::CO, st.carry->getPort(ID::CO));
+
+	auto I2 = st.lut->getPort(ID(I2));
+	if (pm.sigmap(CI) == pm.sigmap(I2)) {
+		cell->setParam(ID(I2_IS_CI), State::S1);
+		I2 = State::Sx;
+	}
+	else
+		cell->setParam(ID(I2_IS_CI), State::S0);
+	cell->setPort(ID(I2), I2);
+	cell->setPort(ID(I3), st.lut->getPort(ID(I3)));
+	cell->setPort(ID::O, st.lut->getPort(ID::O));
+	cell->setParam(ID::LUT, st.lut->getParam(ID(INIT)));
+
+	for (const auto &a : st.carry->attributes)
+		cell->attributes[stringf("\\QL_CARRY.%s", a.first.c_str())] = a.second;
+	for (const auto &a : st.lut->attributes)
+		cell->attributes[stringf("\\LUT4.%s", a.first.c_str())] = a.second;
+	cell->attributes[ID(LUT4.name)] = Const(st.lut->name.str());
+	if (st.carry->get_bool_attribute(ID::keep) || st.lut->get_bool_attribute(ID::keep))
+		cell->attributes[ID::keep] = true;
+
+	pm.autoremove(st.carry);
+	pm.autoremove(st.lut);
+}
+
+struct AP3WrapCarryPass : public Pass {
```

Comment:
Can this be accomplished with techmap pass?  I find techmap passes to be easier to follow than this.

Reply:
I would need to determine what this pass actually does. Possibly consult that with QuickLogic people. I'll take a look on that.

Reply:
@litghost @rw1nkler From what I can tell this pass operates on `LUT4` + `QL_CARRY` pairs. Yosys techmaps can operate on single cells only. A techmap may be used to split one cell into many but not the other way around.

Reply:
Ok